src/devices/src/virtio/block/disk/mod.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2019 The ChromiumOS Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+use std::fs::File;
+use std::io::{self, Read, Seek, SeekFrom};
+
+// QCOW magic constant that starts the header.
+pub const QCOW_MAGIC: u32 = 0x5146_49fb;
+
+#[derive(Debug)]
+pub enum Error {
+    ReadingHeader(io::Error),
+    SeekingFile(io::Error),
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+/// The variants of image files on the host that can be used as virtual disks.
+#[derive(Debug, PartialEq, Eq)]
+pub enum ImageType {
+    Raw,
+    Qcow2,
+}
+/// A trait for getting the length of a disk image or raw block device.
+pub trait DiskGetLen {
+    /// Get the current length of the disk in bytes.
+    fn get_len(&self) -> io::Result<u64>;
+}
+impl DiskGetLen for File {
+    fn get_len(&self) -> io::Result<u64> {
+        let mut s = self;
+        let orig_seek = s.stream_position()?;
+        let end = s.seek(SeekFrom::End(0))?;
+        s.seek(SeekFrom::Start(orig_seek))?;
+        Ok(end)
+    }
+}
+/// Detect the type of an image file by checking for a valid header of the supported formats.
+pub fn detect_image_type(file: &File, overlapped_mode: bool) -> Result<ImageType> {
+    let mut f = file;
+    let disk_size = f.get_len().map_err(Error::SeekingFile)?;
+    let orig_seek = f.stream_position().map_err(Error::SeekingFile)?;
+
+    // Try to read the disk in a nicely-aligned block size unless the whole file is smaller.
+    const MAGIC_BLOCK_SIZE: usize = 4096;
```

Comment:
Is this important? Imago will assume no alignment restrictions unless `direct` is used.

I know it really should probe the actual alignment restrictions (left for a future date…), but also I didn’t know there may be alignment restrictions for non-`direct` I/O. If that’s important, I should fix it. (Either by probing or at least by allowing the “automatically determined” restrictions to be overridden.)

Reply:
I honestly don't know enough to give you a definitive answer. I'm going to look into it.

Reply:
That answer’s good enough for me. :smile: